internal/docstore/drivertest/drivertest.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2019 The Go Cloud Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package drivertest provides a conformance test for implementations of
+// driver.
+package drivertest // import "gocloud.dev/internal/docstore/drivertest"
+
+import (
+	"context"
+	//	"math"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	ds "gocloud.dev/internal/docstore"
+	"gocloud.dev/internal/docstore/driver"
+)
+
+// Harness descibes the functionality test harnesses must provide to run
+// conformance tests.
+type Harness interface {
+	// MakeCollection makes a driver.Collection for testing.
+	MakeCollection(context.Context) (driver.Collection, error)
+
+	// Close closes resources used by the harness.
+	Close()
+}
+
+// HarnessMaker describes functions that construct a harness for running tests.
+// It is called exactly once per test; Harness.Close() will be called when the test is complete.
+type HarnessMaker func(ctx context.Context, t *testing.T) (Harness, error)
+
+// RunConformanceTests runs conformance tests for provider implementations of docstore.
+func RunConformanceTests(t *testing.T, newHarness HarnessMaker) {
+	t.Run("Create", func(t *testing.T) { withCollection(t, newHarness, testCreate) })
+	t.Run("Put", func(t *testing.T) { withCollection(t, newHarness, testPut) })
+	t.Run("Replace", func(t *testing.T) { withCollection(t, newHarness, testReplace) })
+	t.Run("Get", func(t *testing.T) { withCollection(t, newHarness, testGet) })
+	t.Run("Delete", func(t *testing.T) { withCollection(t, newHarness, testDelete) })
+	t.Run("Update", func(t *testing.T) { withCollection(t, newHarness, testUpdate) })
+}
+
+const KeyField = "_id"
+
+func withCollection(t *testing.T, newHarness HarnessMaker, f func(*testing.T, *ds.Collection)) {
+	ctx := context.Background()
+	h, err := newHarness(ctx, t)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer h.Close()
+
+	dc, err := h.MakeCollection(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+	coll := ds.NewCollection(dc)
+	f(t, coll)
+}
+
+type docmap = map[string]interface{}
+
+var nonexistentDoc = docmap{KeyField: "doesNotExist"}
+
+func testCreate(t *testing.T, coll *ds.Collection) {
+	named := docmap{KeyField: "testCreate1", "b": true}
+	unnamed := docmap{"b": false}
+	// Attempt to clean up
+	defer func() {
+		_, _ = coll.Actions().Delete(named).Delete(unnamed).Do(context.Background())
+	}()
+
+	mustRunAndCompare(t, coll, named, coll.Actions().Create(named))
```

Comment:
In tests, I'd rather see similar code repeated in every test than have it factored out like this; it's not obvious what this line is actually doing.

Reply:
Isn't that an argument against every function ever written?

Can we compromise? I'll inline `mustRunAndCompare`, but I think `mustRun` and `shouldFail` are easy to understand and pretty ubiquitous, so I'd like to keep them.

Reply:
Actually, I just defined a few local functions. Let me know if those bother you.